src/middleware.ts

```diff
@@ -116,3 +116,53 @@ export const combine = <
       api as StoreApi<PrimaryState>
     )
   )
+
+export type StateStorage = {
+  getItem: (key: string) => string | null | Promise<string | null>
+  setItem: (key: string, value: string) => void | Promise<void>
+}
+type PersistDefaultOptions<S> = {
+  storage: StateStorage
+  serialize: (state: S) => string | Promise<string>
+  deserialize: (state: string) => S | Promise<S>
+}
+export type PersistOptions<S> = {
+  name: string
+} & Partial<PersistDefaultOptions<S>>
+
+export const persist = <S extends State>(
+  config: StateCreator<S>,
+  options: PersistOptions<S>
+) => (set: SetState<S>, get: GetState<S>, api: StoreApi<S>): S => {
+  const defaultOptions: PersistDefaultOptions<S> = {
+    storage: localStorage,
```

Comment:
This may cause error on non-web environment. Even if we specify `storage` option?

Reply:
Indeed, but you said [this](https://github.com/pmndrs/zustand/pull/229#discussion_r517953668). Am I missing a point?

Reply:
Sorry for the confusion. I didn't quite get the problem then.
Now, I still don't like checking `window`, so my suggestion is this:
```suggestion
    storage: typeof localStorage !== 'undefined' ? localStorage : undefined,
```

But, this can raise an error. Do you have better idea using `typeof localStorage` checking?

Reply:
Do you mean this can raise an error because of `storage` being undefined?

Reply:
Yeah, I mean raising an error is ok because developers need to know that they need to specify it, but error message won't be very friendly.

Reply:
I noticed my suggestion doesn't make sense.
```js
typeof localStorage !== 'undefined' ? localStorage : undefined
```
This is just equivalent to `localStorage`. 😄

Reply:
Testing in chrome console:
<img width="325" alt="image" src="https://user-images.githubusercontent.com/490574/98348758-81b2f080-205c-11eb-8988-dd26aa41e4ac.png">
Okay, it should be good as is.

Reply:
Well, we could throw an error if `defaultOptions.storage && (typeof storage?.getItem !== "function" || typeof storage?.setItem !== "function")` with this `typeof localStorage !== 'undefined' ? localStorage : undefined` (as in your review above) in the default options.

So this would only throw if the code is running in a browser / browser friendly env and if the user has given a custom storage that doesn't has the `getItem` and `setItem` methods.

Reply:
No, I don't think it's necessary. Let's release it and see how people report some issues.

Reply:
On second thought, I think this **is** a problem in RN.
https://github.com/pmndrs/zustand/pull/229/files#r518620500 is working good, but it fails with this.

<img width="454" alt="image" src="https://user-images.githubusercontent.com/490574/98352242-16b7e880-2061-11eb-869a-29433d34648a.png">

So, we need to change the way to pass default options.

Reply:
I think this works:
```js
const {
  name,
  storage = localStorage,
  serialize = JSON.stringify,
  deserialize = JSON.parse,
} = options || {}
```